### PR TITLE
379 support multiple carbon exchanges per chain

### PR DESCRIPTION
**Pull request by mikewcasale**
Handles issue #379

### CODE REVIEW

File: fastlane_bot/tests/test_033_Pools.py [hunk 1]
```diff
@@ -9,10 +9,12 @@
 
 
 import json
+from dataclasses import dataclass
```
Comment (sklbancor) [inline fastlane_bot/tests/test_033_Pools.py:12]: why do you import dataclass yet don't use it?

**Review by sklbancor — CHANGES_REQUESTED** [review]

File: fastlane_bot/bot.py
[the remark this answers is not in the document]
Comment (barakman) [reply fastlane_bot/bot.py:1]: Unrelated of this PR, but we need to move away from these dependencies at some point.
The native types `list`, `dict`, `any`, etc., are all good enough (let alone the fact that we are only really using them in this project as linter hints, so they don't even impact any runtime behavior.

File: fastlane_bot/bot.py [hunk 2]
```diff
@@ -357,9 +357,24 @@ def _convert_trade_instructions(
...
+    def _add_strategy_id_to_trade_instructions_dic(
+        self, trade_instructions_dic: Generator
+    ) -> list:
+        lst = []
+        for ti in trade_instructions_dic:
+            cid = ti["cid"]
+            if '-' in cid:
+                cid = cid.split('-')[0]
+            ti["strategy_id"] = self.db.get_pool(
+                cid=cid
```
Comment (barakman) [inline fastlane_bot/bot.py:373]: Can skip the 'if '-' in cid' above, and just do `cid=cid.split('-')[0]` here.

File: fastlane_bot/helpers/poolandtokens.py [hunk 3]
```diff
@@ -168,6 +169,12 @@ class PoolAndTokens:
 
 
     def __post_init__(self):
+        if not isinstance(self.strategy_id, int):
```
[the remark this answers is not in the document]
Comment (Lesigh-3100) [reply fastlane_bot/helpers/poolandtokens.py:172]: @mikewcasale is there a reason we expect non-integer values for strategy_id here?

File: fastlane_bot/bot.py [hunk 4]
```diff
@@ -357,9 +357,24 @@ def _convert_trade_instructions(
...
+            ).strategy_id
+            lst.append(ti)
+        return lst
```
Comment (barakman) [inline fastlane_bot/bot.py:376]: TBH, I find the cluster of these three functions a little bit hard to follow:
1. `_drop_error`
2. `_convert_trade_instructions`
3. `_add_strategy_id_to_trade_instructions_dic`

And since two of them (1 and 3) are used only in a single place in the code, I suggest merging them into a single function whose code would be IMO easier to follow.

Something like this:
```
def _convert_trade_instructions(self, trade_instructions_dicts):
    errorless_trade_instructions_dicts = [
        {k: v for k, v in trade_instructions_dicts[i].items() if k != "error"}
        for i in range(len(trade_instructions_dicts))
    ]

    tis = [
        {
            **ti,
            "raw_txs": "[]",
            "pair_sorting": "",
            "ConfigObj": self.ConfigObj,
            "db": self.db,
            "strategy_id": self.db.get_pool(cid=ti["cid"].split('-')[0]).strategy_id
        }
        for ti in errorless_trade_instructions_dicts
        if ti is not None
    ]

    return [TradeInstruction(**ti) for ti in tis]
```

**Review by Lesigh-3100 — CHANGES_REQUESTED** [review]
run_blockchain_terraformer.py should set "strategy_id" field for static data if this is an expected field for all pools.

File: fastlane_bot/bot.py [hunk 5]
```diff
@@ -949,6 +965,8 @@ def _handle_trade_instructions(
             trade_instructions=ordered_trade_instructions_objects
         )
 
+        # ordered_trade_instructions_objects[0].db = QueryInterface(ConfigObj=self.ConfigObj)
+
```
Comment (barakman) [inline fastlane_bot/bot.py:969]: Seems like a redundant change

File: fastlane_bot/config/network.py [hunk 6]
```diff
@@ -208,7 +208,8 @@ class ConfigNetwork(ConfigBase):
     UNISWAP_V3_NAME = "uniswap_v3"
```
[the remark this answers is not in the document]
Comment (barakman) [reply fastlane_bot/config/network.py:208]: This has been addressed several times before (I personally asked the same question on a previous PR).
There's probably an issue on this, which can be handled separately of this PR.

File: fastlane_bot/events/async_event_update_utils.py [hunk 7]
```diff
@@ -317,12 +306,16 @@ def process_contract_chunks(
...
 ) -> pd.DataFrame:
     lst = []
     # write chunks to csv
     for idx, chunk in enumerate(chunks):
         loop = asyncio.get_event_loop()
-        df = loop.run_until_complete(func(chunk))
+        if filename == "tokens_and_fee_df.csv":
+            df = loop.run_until_complete(func(chunk, carbon_v1_forks))
+        else:
+            df = loop.run_until_complete(func(chunk))
```
Comment (barakman) [inline fastlane_bot/events/async_event_update_utils.py:318]: I see no advantage in passing `func` as input to this function, so I recommend just calling the related functions directly (`main_get_tokens_and_fee` in the 'if' and `main_get_missing_tkn` in the 'else').

If these two functions had an identical interface (input and output), then there would be some advantage in doing so (but even under such scenario, the filename-based decision on which function to call, would still reduce that advantage to a minimum).

Side note: since both of these functions are used only in this module, you can also prepend the name of each one of them with an underscore.

**Review by barakman — CHANGES_REQUESTED** [review]

File: fastlane_bot/config/network.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sklbancor) [reply fastlane_bot/config/network.py:208]: how can we collect those issues that are not bad enough to prevent merging but that should need to be addressed?

File: fastlane_bot/helpers/routehandler.py [hunk 8]
```diff
@@ -1674,6 +1678,7 @@ def calculate_trade_outputs(
         return trade_instructions
 
     def _from_wei_to_decimals(self, tkn0_amt: Decimal, tkn0_decimals: int) -> Decimal:
+        print(f"[_from_wei_to_decimals] tkn0_amt={tkn0_amt}, tkn0_decimals={tkn0_decimals}")
```
[the remark this answers is not in the document]
Comment (sklbancor) [reply fastlane_bot/helpers/routehandler.py:1681]: how hard would it be to make a Github action that checks whether there are not-commented-out print statements in the code and possibly prevent merging if this is the case?

File: fastlane_bot/events/pools/carbon_v1.py [hunk 9]
```diff
@@ -12,15 +12,16 @@
...
 
 
 @dataclass
 class CarbonV1Pool(Pool):
     """
     Class representing a Carbon v1 pool.
     """
-
-    exchange_name: str = "carbon_v1"
+    base_exchange_name: str = "carbon_v1"
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/pools/carbon_v1.py:23]: This class is not duplicated, and not sure what comment above you're referring to.

File: fastlane_bot/events/pools/carbon_v1.py [hunk 10]
```diff
@@ -51,7 +52,7 @@ def update_from_event(
             "This event should not be " "handled by this class."
         )
         data = CarbonV1Pool.parse_event(data, event_args, event_type)
-        data["router"] = self.router_address,
+        data["router"] = self.router_address
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/pools/carbon_v1.py:55]: Somewhat shockingly, this actually does not throw an error. We do have tests that would catch it if it did throw an error, but Ive confirmed the tests pass, and even wrote a simple proof to confirm it. But anyway good to fix it now

image

File: fastlane_bot/bot.py
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/bot.py:1]: I have corrected this

File: fastlane_bot/config/network.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/config/network.py:208]: I have added a new issue for this here: https://github.com/bancorprotocol/fastlane-bot/issues/456

File: fastlane_bot/config/selectors.py
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/config/selectors.py:1]: It is not needed, and also should be handled in this issue: https://github.com/bancorprotocol/fastlane-bot/issues/456

File: fastlane_bot/events/async_event_update_utils.py [hunk 11]
```diff
@@ -79,40 +80,30 @@ async def get_token_and_fee(
...
+            "strategy_id": strategy_id
+        }
+        cid = get_pool_cid(pool_info, carbon_v1_forks=carbon_v1_forks)
+        return exchange_name, address, tkn0, tkn1, fee, cid, strategy_id, anchor
     except Exception as e:
         cfg.logger.info(
             f"Failed to get tokens and fee for {address} {exchange_name} {e}"
         )
-        return exchange_name, address, None, None, None, None, anchor
+        return exchange_name, address, None, None, None, None, None, anchor
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/async_event_update_utils.py]: I have corrected this

File: fastlane_bot/events/async_event_update_utils.py [hunk 12]
```diff
@@ -166,16 +158,13 @@ def get_pool_info(
         "tkn0_decimals": tkn0["decimals"],
         "tkn1_symbol": tkn1["symbol"],
         "tkn1_decimals": tkn1["decimals"],
-        "pair_name": tkn0["address"] + "/" + tkn1["address"]
+        "pair_name": tkn0["address"] + "/" + tkn1["address"],
+        "strategy_id": pool["strategy_id"] if "strategy_id" in pool else 0,
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/async_event_update_utils.py]: I have corrected this

File: fastlane_bot/events/async_event_update_utils.py [hunk 13]
```diff
@@ -317,12 +306,16 @@ def process_contract_chunks(
         func: Callable,
         df_combined: pd.DataFrame = None,
         read_only: bool = False,
+        carbon_v1_forks: List[str] = None,
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/async_event_update_utils.py]: I have corrected this

File: fastlane_bot/events/async_event_update_utils.py [hunk 14]
```diff
@@ -317,12 +306,16 @@ def process_contract_chunks(
...
 ) -> pd.DataFrame:
     lst = []
     # write chunks to csv
     for idx, chunk in enumerate(chunks):
         loop = asyncio.get_event_loop()
-        df = loop.run_until_complete(func(chunk))
+        if filename == "tokens_and_fee_df.csv":
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/async_event_update_utils.py:376]: I have added a new issue for this https://github.com/bancorprotocol/fastlane-bot/issues/462

File: fastlane_bot/events/exchanges/carbon_v1.py [hunk 15]
```diff
@@ -15,15 +15,16 @@
...
 
 
 @dataclass
 class CarbonV1(Exchange):
     """
     Carbon exchange class
     """
-
-    exchange_name: str = "carbon_v1"
+    base_exchange_name: str = "carbon_v1"
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/exchanges/carbon_v1.py:26]: The current implementation (not ideal but...) handles instances of forks of an exchange as if they are the original exchange with a few exceptions which are checked for via something along the lines of `is_exchange_fork = exchange_name != base_exchange_name`. Thus, is_exchange_fork=False when the CarbonV1Exchange is the original carbon deployment, and True if it is a fork thereof.

File: fastlane_bot/events/managers/manager.py [hunk 16]
```diff
@@ -18,43 +18,40 @@
 
 
 class Manager(PoolManager, EventManager, ContractsManager):
-    def update_from_event(self, event: Dict[str, Any]) -> None:
+    def update_from_event(self, event: Dict[str, Any]):
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/managers/manager.py:21]: Not sure what exactly you're referring to here

File: fastlane_bot/events/new_utils.py [hunk 17]
```diff
@@ -0,0 +1,52 @@
+from typing import Any, Dict, List
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/new_utils.py]: this is fixed

File: fastlane_bot/events/new_utils.py [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+
+from web3 import Web3
+
+
+def get_pool_cid(args: Dict[str, Any], carbon_v1_forks: List[str]) -> str:
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/new_utils.py]: You can use lowercase `str` type-hints with python 3.8, similar to `int`. It's only the others (list, dict,tuple, etc)

File: fastlane_bot/events/new_utils.py [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+    """
+    Get the pool CID from the event args
+
+    Parameters
+    ----------
+    args : Dict[str, Any]
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/new_utils.py]: this is now changed to Google style, thanks. (and this should always be the standard from now on)

File: fastlane_bot/events/pools/carbon_v1.py [hunk 20]
```diff
@@ -153,7 +154,13 @@ def update_from_contract(
         params = self.parse_event(self.state, fake_event, "None")
         params["exchange_name"] = self.exchange_name
         params["router"] = self.router_address,
+        params["fee"] = self.state["fee"]
+        fake_carbon_v1_fork_list = [self.exchange_name]  # because we dont have access to the config object from here,
```
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/events/pools/carbon_v1.py]: This has been corrected

File: fastlane_bot/helpers/poolandtokens.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/helpers/poolandtokens.py]: This has been corrected (i.e. removed altogether). no reason to expect it as anything other than int

File: fastlane_bot/bot.py (hunk 2 above)
Comment (mikewcasale) [reply fastlane_bot/bot.py:367]: this has been corrected

File: fastlane_bot/bot.py (hunk 4 above)
Comment (mikewcasale) [reply fastlane_bot/bot.py:370]: this has been corrected

File: fastlane_bot/bot.py (hunk 5 above)
Comment (mikewcasale) [reply fastlane_bot/bot.py:963]: not sure what you're referring to here

File: fastlane_bot/events/async_event_update_utils.py (hunk 7 above)
Comment (mikewcasale) [reply fastlane_bot/events/async_event_update_utils.py:377]: The original interface was the same, and thanks to my most recent changes they are now the same again, so this has been corrected

File: fastlane_bot/data/multichain_addresses.csv
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/data/multichain_addresses.csv:1]: I have talked with Nick about this and unless you have objections, we're keeping the 'altered_carbon' name throughout

File: fastlane_bot/config/constants.py
[the remark this answers is not in the document]
Comment (mikewcasale) [reply fastlane_bot/config/constants.py:1]: I have talked with Nick about this and unless you have objections, we're keeping the 'altered_carbon' name throughout. Thus, I removed the comment about changing it before merging

File: fastlane_bot/data/multichain_addresses.csv
[the remark this answers is not in the document]
Comment (sklbancor) [reply fastlane_bot/data/multichain_addresses.csv:1]: I have no objections, but then remove the comment that says "change before release"; resolve when done

File: fastlane_bot/events/pools/carbon_v1.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (sklbancor) [reply fastlane_bot/events/pools/carbon_v1.py:55]: `x = 1,` is not illegal; it is short for `x=(1,)` so the result is a tuple so somehow this value never made it anywhere

File: fastlane_bot/events/exchanges/carbon_v1.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (sklbancor) [reply fastlane_bot/events/exchanges/carbon_v1.py:25]: I created an issue for this

File: fastlane_bot/events/new_utils.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (sklbancor) [reply fastlane_bot/events/new_utils.py]: OK

File: fastlane_bot/bot.py (hunk 5 above)
Comment (barakman) [reply fastlane_bot/bot.py:963]: This is new code (painted green), right?
And it is masked out, right?
So that makes it a redundant change as far as I can say (did you forget to remove it?).

File: fastlane_bot/bot.py [hunk 21]
```diff
@@ -354,12 +350,25 @@ def _convert_trade_instructions(
                 "ConfigObj": self.ConfigObj,
                 "db": self.db,
```
Comment (barakman) [inline fastlane_bot/bot.py:351]: Just in case you missed that part in my previous suggestion, you can add here:
```
"strategy_id": self.db.get_pool(cid=ti["cid"].split('-')[0]).strategy_id
```
Then get rid of function `_add_strategy_id_to_trade_instructions_dic`.
It would be overall cleaner IMO.
Feel free to resolve this comment regardless of your decision :)

File: fastlane_bot/helpers/routehandler.py [hunk 22]
```diff
@@ -1528,7 +1530,7 @@ def calculate_trade_outputs(
         List[Dict[str, Any]]
             The trade outputs.
         """
-
+        db = trade_instructions[0].db
```
Comment (barakman) [inline fastlane_bot/helpers/routehandler.py:1533]: Why is this necessary?

**Review by barakman — CHANGES_REQUESTED** [review]

File: fastlane_bot/helpers/routehandler.py (hunk 22 above)
Comment (mikewcasale) [reply fastlane_bot/helpers/routehandler.py:1533]: pretty sure I have not touched this, it's just a deleted empty line

File: fastlane_bot/events/managers/base.py [hunk 23]
```diff
@@ -172,18 +172,31 @@ def set_carbon_v1_fee_pairs(self):
         """
         Set the carbon v1 fee pairs.
         """
-        if "carbon_v1" in self.exchanges:
```
Comment (barakman) [inline fastlane_bot/events/managers/base.py:175]: I think that by changing this `if` statement into:
```
for ex in [ex for ex in self.cfg.CARBON_V1_FORKS if ex in self.exchanges]:
```
You could keep most of the code below exactly as is.
Fewer changes == less risk AND better review.

File: fastlane_bot/events/managers/base.py [hunk 24]
```diff
@@ -238,6 +251,11 @@ def exchange_name_from_event(self, event: Dict[str, Any]) -> str:
         str
             The exchange name.
         """
+        if 'id' in event['args']:
+            carbon_controller_address = event['address']
+            for ex in self.cfg.CARBON_CONTROLLER_MAPPING:
```
Comment (barakman) [inline fastlane_bot/events/managers/base.py:256]: Semantics: probably better using `exchange_name` to keep it consistent with the rest of this function.
And it's probably true for a bunch of other places below and above (wherever you're returning the name of the exchange and not the exchange object).

File: fastlane_bot/helpers/routehandler.py (hunk 22 above)
Comment (barakman) [reply fastlane_bot/helpers/routehandler.py:1533]: I see a green line with `db = trade_instructions[0].db`, and nowhere below do I see `db` being used.

And even if it is used somewhere and I've missed that somehow, this usage is not marked as a change, which means that up until now, there was an attempt in the code to use a non-existing variable, which would have caused an immediate exception; is that really the case here (i.e., does that green line fix a bug)?

**Review by barakman — CHANGES_REQUESTED** [review]

File: fastlane_bot/bot.py (hunk 5 above)
Comment (mikewcasale) [reply fastlane_bot/bot.py]: got it. fixed, thanks

File: fastlane_bot/helpers/routehandler.py (hunk 22 above)
Comment (mikewcasale) [reply fastlane_bot/helpers/routehandler.py]: As far as I can see, you're correct that the db variable was never used. While I dont recall touching this code, perhaps Github has a better memory than me... in either case, I have now removed the line, which I believe resolves this conversation

File: fastlane_bot/events/managers/base.py (hunk 23 above)
Comment (mikewcasale) [reply fastlane_bot/events/managers/base.py:175]: I have implemented your suggestion. thanks

File: fastlane_bot/events/managers/base.py (hunk 24 above)
Comment (mikewcasale) [reply fastlane_bot/events/managers/base.py:256]: due to the number of related changes involved with handling every similar occurrence, I have created a new issue for this https://github.com/bancorprotocol/fastlane-bot/issues/469

**Review by barakman — DISMISSED** [review]

**Review by sklbancor — DISMISSED** [review]

**Review by Lesigh-3100 — APPROVED** [review]

**Review by zavelevsky — DISMISSED** [review]
LGTM

File: fastlane_bot/events/async_event_update_utils.py [hunk 25]
```diff
@@ -79,57 +148,42 @@ async def get_token_and_fee(
...
+        pool_info = {
+            "exchange_name": exchange_name,
+            "address": address,
+            "tkn0_address": tkn0,
+            "tkn1_address": tkn1,
+            "pair_name": f"{tkn0}/{tkn1}",
+            "fee": fee,
+            "strategy_id": strategy_id
+        }
+        carbon_v1_forks = [CARBON_V1_NAME] if ex.is_carbon_v1_fork else []
```
Comment (barakman) [inline fastlane_bot/events/async_event_update_utils.py:163]: Shouldn't this be chain-dependent (or passed as input from the caller, ultimately depending on the chain)?
Otherwise, what exactly has been achieved here, as a single carbon deployment is eventually handled?

File: fastlane_bot/events/async_event_update_utils.py [hunk 26]
```diff
@@ -166,16 +220,13 @@ def get_pool_info(
         "tkn0_decimals": tkn0["decimals"],
         "tkn1_symbol": tkn1["symbol"],
         "tkn1_decimals": tkn1["decimals"],
-        "pair_name": tkn0["address"] + "/" + tkn1["address"]
+        "pair_name": tkn0["address"] + "/" + tkn1["address"],
```
Comment (barakman) [inline fastlane_bot/events/async_event_update_utils.py:223]: Minor: since you're using `pool_info["pair_name"].split("/")` below...

...you can also use `"/".join(tkn0["address"], tkn1["address"])` here.

**Review by barakman — CHANGES_REQUESTED** [review]

File: fastlane_bot/events/async_event_update_utils.py (hunk 25 above)
Comment (mikewcasale) [reply fastlane_bot/events/async_event_update_utils.py]: oh man, good catch! I just pushed a fix

File: fastlane_bot/events/async_event_update_utils.py (hunk 26 above)
Comment (mikewcasale) [reply fastlane_bot/events/async_event_update_utils.py:223]: Im not a big fan of the join method as I find it less readable

File: fastlane_bot/events/async_event_update_utils.py (hunk 25 above)
Comment (barakman) [reply fastlane_bot/events/async_event_update_utils.py]: Sure thing man, but doesn't that still yield a list of one item, i.e., a single carbon deployment?

Comment (mikewcasale) [reply fastlane_bot/events/async_event_update_utils.py]: Yes but in this case that is the intent. If you look at the function to get the cid, all we need is for the exchange_name to exist in the list passed to that function - typically this is an instance of the config object, but we dont actually need the config object in this case because if the instance of the exchange object is either a fork of carbon, or carbon, then this satisfies the purpose of checking the list of carbon forks (ie we can be sure that it is indeed a carbon fork without needing the config object).

**Review by barakman — APPROVED** [review]
